tests/lib/nrf_modem_lib/nrf_modem_lib_trace/src/main.c

```diff
@@ -0,0 +1,329 @@
+static uint32_t trace_processed_callback_len_expect;
+static int trace_processed_callback_retval;
+
+static int trace_processed_callback_stub(const uint8_t *buf, uint32_t len, int cmock_calls)
+{
+	TEST_ASSERT_EQUAL(buf, trace_processed_callback_buf_expect);
+	TEST_ASSERT_EQUAL(len, trace_processed_callback_len_expect);
+
+	return trace_processed_callback_retval;
+}
+
+/* Custom mock written for nrf_modem_trace_processed_callback since the regular mock
+ * function, __wrap_nrf_modem_trace_processed_callback_ExpectAndReturn, erroneously
+ * calls the fault handler when compiler is optimizing for size.
+ */
+static void nrf_modem_trace_processed_callback_ExpectAndReturn(const uint8_t *buf,
+							       uint32_t len, int retval)
+{
+	trace_processed_callback_buf_expect = buf;
+	trace_processed_callback_len_expect = len;
+	trace_processed_callback_retval = retval;
+
+	__wrap_nrf_modem_trace_processed_callback_Stub(trace_processed_callback_stub);
+}
+
+int test_suiteTearDown(int num_failures)
+{
+	return generic_suiteTearDown(num_failures);
+}
+
+/* Test that nrf_modem_lib_trace_start returns error when modem trace module is not initialized. */
+void test_nrf_modem_trace_start_eperm(void)
+{
+	int ret;
+
+	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL);
+
+	TEST_ASSERT_EQUAL(-EPERM, ret);
+}
+
+void test_nrf_modem_lib_trace_init(void)
+{
+	int ret;
+
+	__wrap_trace_backend_init_ExpectAndReturn(0);
+
+	ret = nrf_modem_lib_trace_init(&trace_heap);
+
+	TEST_ASSERT_EQUAL(0, ret);
+}
+
+void test_nrf_modem_lib_trace_start_coredump_only(void)
+{
+	int ret;
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_COREDUMP_ONLY, 0);
+
+	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_COREDUMP_ONLY);
+
+	TEST_ASSERT_EQUAL(0, ret);
+}
+
+void test_nrf_modem_lib_trace_start_all(void)
+{
+	int ret;
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_ALL, 0);
+
+	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL);
+
+	TEST_ASSERT_EQUAL(0, ret);
+}
+
+void test_nrf_modem_lib_trace_start_ip_only(void)
+{
+	int ret;
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_IP_ONLY, 0);
+
+	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_IP_ONLY);
+
+	TEST_ASSERT_EQUAL(0, ret);
+}
+
+void test_nrf_modem_lib_trace_start_lte_ip(void)
+{
+	int ret;
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_LTE_IP, 0);
+
+	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_LTE_IP);
+
+	TEST_ASSERT_EQUAL(0, ret);
+}
+
+/* Test that traces are forwarded to the trace backend when the module is correctly initialized
+ * and nrf_modem_lib_trace_start has been called.
+ */
+void test_modem_trace_write_to_backend(void)
+{
+	int ret;
+	const uint8_t sample_trace_data[10];
+
+	test_nrf_modem_lib_trace_init();
+
+	nrf_modem_at_printf_ExpectTraceModeAndReturn(NRF_MODEM_LIB_TRACE_ALL, 0);
+	TEST_ASSERT_EQUAL(0, nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL));
```

Comment:
If we want to follow this pattern. Applies for other functions as well.
```suggestion
	ret = nrf_modem_lib_trace_start(NRF_MODEM_LIB_TRACE_ALL);
	TEST_ASSERT_EQUAL(0, ret);
```

Reply:
Thanks, I missed a couple of those